### PR TITLE
Add recordLogs, and matchers to process the results.

### LINKED ISSUE #372: Expand "scopeLog" to make testing *way* easier
Today I need the following boilerplate (from an AngularDart test):

```dart
    test('a class with an invalid field formal parameter', () async {
      final logs = <String>[];
      final logger = new Logger.detached('test')..onRecord.listen((r) {
        print(r);
        logs.add(r.message);
      });
      scopeLog(() {
        reader.parseDependencies(classNamed('BadField'));
      }, logger);
      await new Future.value();
      expect(logs, ['Bad thing happened']);
    });
```

It would be nice to reduce this, somehow, to:

```dart
    test('a class with an invalid field formal parameter', () async {
      final logs = await scopeTestLogger(() {
        reader.parseDependencies(classNamed('BadField'));
      });
      expect(logs, ['Bad thing happened']);
    });
```

### PR DESCRIPTION
**Pull request by matanlurey**
Closes #372.

I could have gone all in on different matcher combinations, but I'd rather wait for the need.

/cc @kevmoo

### CODE REVIEW

File: build_test/lib/src/matchers.dart [hunk 1]
```diff
@@ -5,10 +5,18 @@ import 'package:test/test.dart';
 
 import 'package:build/build.dart';
 
+/// Matches that the provided instance is a [AssetNotFoundException].
```
Comment (natebosch) [inline build_test/lib/src/matchers.dart]: [nit] This reads strangely to me. Maybe "Matches instances of [AssetNotFoundException]."

File: build_test/lib/src/record_logs.dart [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:build/src/builder/logging.dart';
+import 'package:matcher/matcher.dart';
+import 'package:logging/logging.dart';
+
+/// Executes [run] with a new [Logger], returning the resulting log records.
```
Comment (natebosch) [inline build_test/lib/src/record_logs.dart]: Can you mention how this behaves when `run` does or doesn't return a Future?

Also mention that the returned Stream never closes.

File: build_test/lib/src/record_logs.dart [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    _message.describe(description);
+    return description;
+  }
+
+  @override
+  Description describeMismatch(
+    covariant LogRecord item,
+    Description description,
+    _,
+    __,
```
Comment (jakemac53) [inline build_test/lib/src/record_logs.dart]: nit: remove trailing comma :)

File: build_test/lib/src/record_logs.dart [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  return logger.onRecord;
+}
+
+/// Matches [LogRecord] of any level whose message is [messageOrMatcher].
+///
+/// ```dart
+/// anyLogOf('Hello World)';     // Exactly match 'Hello World'.
+/// anyLogOf(contains('ERROR')); // Contains the sub-string 'ERROR'.
+/// ```
+Matcher anyLogOf(dynamic messageOrMatcher) =>
```
Comment (jakemac53) [inline build_test/lib/src/record_logs.dart]: I could imagine `greaterThan` and `lessThan` optional args here (would be `Level`s), but probably not really worth the complexity.

File: build_test/lib/src/matchers.dart (hunk 1 above)
Comment (matanlurey) [reply build_test/lib/src/matchers.dart]: Done.

File: build_test/lib/src/record_logs.dart (hunk 2 above)
Comment (matanlurey) [reply build_test/lib/src/record_logs.dart]: Done.

File: build_test/lib/src/record_logs.dart (hunk 4 above)
Comment (matanlurey) [reply build_test/lib/src/record_logs.dart]: Yeah. I'll wait until needed ;)

File: build_test/lib/src/record_logs.dart (hunk 3 above)
Comment (matanlurey) [reply build_test/lib/src/record_logs.dart]: Done.

**Pull request merged**